File: DiKErnel/Test/DiKErnel.DomainLibrary.Test/Validators/GrassRevetmentWaveImpact/GrassRevetmentWaveImpactValidatorTest.cpp
// Copyright (C) Stichting Deltares 2020. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, version 3.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

#include <limits>

#include <gtest/gtest.h>

#include "GrassRevetmentWaveImpactValidator.h"
#include "ValidatorAssertHelper.h"

namespace DiKErnel::DomainLibrary::Test
{
    using namespace std;
    using namespace testing;
    using namespace TestUtil;

    struct GrassRevetmentWaveImpactValidatorTest : Test
    {
        static unique_ptr<ValidationIssue> TimeLineAgwi(
            const double timeLineAgwi)
        {
            return GrassRevetmentWaveImpactValidator::TimeLineAgwi(timeLineAgwi, 3);
        }

        static unique_ptr<ValidationIssue> UpperLimitLoadingAul(
            const double upperLimitLoadingAul)
        {
            return GrassRevetmentWaveImpactValidator::UpperLimitLoadingAul(upperLimitLoadingAul, 2);
        }
    };

    TEST_F(GrassRevetmentWaveImpactValidatorTest, TimeLineAgwi_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = TimeLineAgwi;

        constexpr auto errorMessage = "TimeLineAgwi must be larger than TimeLineCgwi.";

        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, ValidatorAssertHelper::DOUBLE_MIN, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, 3.0 - ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, 3.0, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, 3.0 + ValidatorAssertHelper::EPSILON);

        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, ValidatorAssertHelper::DOUBLE_MAX);
    }

    TEST_F(GrassRevetmentWaveImpactValidatorTest, TimeLineB1gwi_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = GrassRevetmentWaveImpactValidator::TimeLineBgwi;

        constexpr auto errorMessage = "TimeLineBgwi must be smaller than 0";

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MIN);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0 - ValidatorAssertHelper::EPSILON);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0 + ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MAX, ValidationIssueType::Error, errorMessage);
    }

    TEST_F(GrassRevetmentWaveImpactValidatorTest, TimeLineCgwi_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = GrassRevetmentWaveImpactValidator::TimeLineCgwi;

        constexpr auto errorMessage = "TimeLineCgwi must be equal to 0 or larger.";

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MIN, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0 - ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0 + ValidatorAssertHelper::EPSILON);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MAX);
    }

    TEST_F(GrassRevetmentWaveImpactValidatorTest, MinimumWaveHeightTemax_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = GrassRevetmentWaveImpactValidator::MinimumWaveHeightTemax;

        constexpr auto errorMessage = "MinimumWaveHeightTemax must be equal to 3600000 or smaller.";
        constexpr auto warningMessage = "MinimumWaveHeightTemax should be in range {1000000, 3600000].";

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MIN, ValidationIssueType::Warning, warningMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1000000.0 - ValidatorAssertHelper::EPSILON, ValidationIssueType::Warning,
                                                  warningMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1000000.0, ValidationIssueType::Warning, warningMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1000000.0 + ValidatorAssertHelper::EPSILON);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 3600000.0 - ValidatorAssertHelper::EPSILON);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 3600000.0);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 3600000.0 + ValidatorAssertHelper::EPSILON, ValidationIssueType::Error,
                                                  errorMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MAX, ValidationIssueType::Error, errorMessage);
    }

    TEST_F(GrassRevetmentWaveImpactValidatorTest, MaximumWaveHeightTemin_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = GrassRevetmentWaveImpactValidator::MaximumWaveHeightTemin;

        constexpr auto errorMessage = "MaximumWaveHeightTemin must be equal to 3.6 or larger.";
        constexpr auto warningMessage = "MaximumWaveHeightTemin should be in range [3.6, 10}.";

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MIN, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 3.6 - ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 3.6);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 3.6 + ValidatorAssertHelper::EPSILON);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 10.0 - ValidatorAssertHelper::EPSILON);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 10.0, ValidationIssueType::Warning, warningMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 10.0 + ValidatorAssertHelper::EPSILON, ValidationIssueType::Warning,
                                                  warningMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MAX, ValidationIssueType::Warning, warningMessage);
    }

    TEST_F(GrassRevetmentWaveImpactValidatorTest, WaveAngleImpactNwa_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = GrassRevetmentWaveImpactValidator::WaveAngleImpactNwa;

        constexpr auto errorMessage = "WaveAngleImpactNwa must be equal to 1 or smaller.";

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MIN);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1.0 - ValidatorAssertHelper::EPSILON);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1.0);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1.0 + ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MAX, ValidationIssueType::Error, errorMessage);
    }

    TEST_F(GrassRevetmentWaveImpactValidatorTest, WaveAngleImpactQwa_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = GrassRevetmentWaveImpactValidator::WaveAngleImpactQwa;

        constexpr auto errorMessage = "WaveAngleImpactQwa must be in range [0, 1].";

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MIN, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0 - ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0 + ValidatorAssertHelper::EPSILON);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1.0 - ValidatorAssertHelper::EPSILON);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1.0);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 1.0 + ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MAX, ValidationIssueType::Error, errorMessage);
    }

    TEST_F(GrassRevetmentWaveImpactValidatorTest, WaveAngleImpactRwa_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = GrassRevetmentWaveImpactValidator::WaveAngleImpactRwa;

        constexpr auto errorMessage = "WaveAngleImpactRwa must be larger than 0.";

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MIN, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0 - ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound(validateAction, 0.0 + ValidatorAssertHelper::EPSILON);

        ValidatorAssertHelper::AssertEqualToBound(validateAction, ValidatorAssertHelper::DOUBLE_MAX);
    }

    TEST_F(GrassRevetmentWaveImpactValidatorTest, UpperLimitLoadingAul_VariousScenarios_ExpectedValues)
    {
        const auto validateAction = UpperLimitLoadingAul;

        constexpr auto errorMessage = "UpperLimitLoadingAul must be smaller than LowerLimitLoadingAll.";

        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, ValidatorAssertHelper::DOUBLE_MIN);

        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, 2.0 - ValidatorAssertHelper::EPSILON);
        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, 2.0, ValidationIssueType::Error, errorMessage);
        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, 2.0 + ValidatorAssertHelper::EPSILON, ValidationIssueType::Error, errorMessage);

        ValidatorAssertHelper::AssertEqualToBound<double>(validateAction, ValidatorAssertHelper::DOUBLE_MAX, ValidationIssueType::Error, errorMessage);
    }
}

File: Test/DiKErnel.DomainLibrary.Test/Validators/GrassRevetmentWaveImpact/GrassRevetmentWaveImpactValidatorTest.cs
// Copyright (C) Stichting Deltares and State of the Netherlands 2023. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify it under the terms of the
// GNU Lesser General Public License as published by the Free Software Foundation, either
// version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License along with this
// program. If not, see <http://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of Stichting
// Deltares and remain full property of Stichting Deltares at all times. All rights reserved.

using System;
using DiKErnel.DomainLibrary.Validators.GrassRevetmentWaveImpact;
using DiKErnel.Util.TestUtil;
using DiKErnel.Util.Validation;
using NUnit.Framework;

namespace DiKErnel.DomainLibrary.Test.Validators.GrassRevetmentWaveImpact
{
    [TestFixture]
    public class GrassRevetmentWaveImpactValidatorTest
    {
        [Test]
        public void TimeLineAgwi_VariousScenarios_ExpectedValues()
        {
            ValidationIssue ValidationFunc(double timeLineAgwi) => GrassRevetmentWaveImpactValidator.TimeLineAgwi(timeLineAgwi, 3);

            const string errorMessage = "TimeLineAgwi must be larger than TimeLineCgwi.";

            ValidatorAssertHelper.AssertValue(ValidationFunc, ValidatorAssertHelper.DoubleMin, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(ValidationFunc, 3.0 - ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(ValidationFunc, 3.0, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(ValidationFunc, 3.0 + ValidatorAssertHelper.Epsilon);

            ValidatorAssertHelper.AssertValue(ValidationFunc, ValidatorAssertHelper.DoubleMax);
        }

        [Test]
        public void TimeLineBgwi_VariousScenarios_ExpectedValues()
        {
            Func<double, ValidationIssue> validationFunc = GrassRevetmentWaveImpactValidator.TimeLineBgwi;

            const string errorMessage = "TimeLineBgwi must be smaller than 0.";

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMin);

            ValidatorAssertHelper.AssertValue(validationFunc, 0.0 - ValidatorAssertHelper.Epsilon);
            ValidatorAssertHelper.AssertValue(validationFunc, 0.0, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 0.0 + ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMax, ValidationIssueType.Error, errorMessage);
        }

        [Test]
        public void TimeLineCgwi_VariousScenarios_ExpectedValues()
        {
            Func<double, ValidationIssue> validationFunc = GrassRevetmentWaveImpactValidator.TimeLineCgwi;

            const string errorMessage = "TimeLineCgwi must be equal to 0 or larger.";

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMin, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, 0.0 - ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 0.0);
            ValidatorAssertHelper.AssertValue(validationFunc, 0.0 + ValidatorAssertHelper.Epsilon);

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMax);
        }

        [Test]
        public void MinimumWaveHeightTemax_VariousScenarios_ExpectedValues()
        {
            Func<double, ValidationIssue> validationFunc = GrassRevetmentWaveImpactValidator.MinimumWaveHeightTemax;

            const string errorMessage = "MinimumWaveHeightTemax must be equal to 3600000 or smaller.";
            const string warningMessage = "MinimumWaveHeightTemax should be in range {1000000, 3600000].";

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMin, ValidationIssueType.Warning, warningMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, 1000000.0 - ValidatorAssertHelper.Epsilon, ValidationIssueType.Warning,
                                              warningMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 1000000.0, ValidationIssueType.Warning, warningMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 1000000.0 + ValidatorAssertHelper.Epsilon);

            ValidatorAssertHelper.AssertValue(validationFunc, 3600000.0 - ValidatorAssertHelper.Epsilon);
            ValidatorAssertHelper.AssertValue(validationFunc, 3600000.0);
            ValidatorAssertHelper.AssertValue(validationFunc, 3600000.0 + ValidatorAssertHelper.Epsilon, ValidationIssueType.Error,
                                              errorMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMax, ValidationIssueType.Error, errorMessage);
        }

        [Test]
        public void MaximumWaveHeightTemin_VariousScenarios_ExpectedValues()
        {
            Func<double, ValidationIssue> validationFunc = GrassRevetmentWaveImpactValidator.MaximumWaveHeightTemin;

            const string errorMessage = "MaximumWaveHeightTemin must be equal to 3.6 or larger.";
            const string warningMessage = "MaximumWaveHeightTemin should be in range [3.6, 10}.";

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMin, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, 3.6 - ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 3.6);
            ValidatorAssertHelper.AssertValue(validationFunc, 3.6 + ValidatorAssertHelper.Epsilon);

            ValidatorAssertHelper.AssertValue(validationFunc, 10.0 - ValidatorAssertHelper.Epsilon);
            ValidatorAssertHelper.AssertValue(validationFunc, 10.0, ValidationIssueType.Warning, warningMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 10.0 + ValidatorAssertHelper.Epsilon, ValidationIssueType.Warning,
                                              warningMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMax, ValidationIssueType.Warning, warningMessage);
        }

        [Test]
        public void WaveAngleImpactNwa_VariousScenarios_ExpectedValues()
        {
            Func<double, ValidationIssue> validationFunc = GrassRevetmentWaveImpactValidator.WaveAngleImpactNwa;

            const string errorMessage = "WaveAngleImpactNwa must be equal to 1 or smaller.";

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMin);

            ValidatorAssertHelper.AssertValue(validationFunc, 1.0 - ValidatorAssertHelper.Epsilon);
            ValidatorAssertHelper.AssertValue(validationFunc, 1.0);
            ValidatorAssertHelper.AssertValue(validationFunc, 1.0 + ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMax, ValidationIssueType.Error, errorMessage);
        }

        [Test]
        public void WaveAngleImpactQwa_VariousScenarios_ExpectedValues()
        {
            Func<double, ValidationIssue> validationFunc = GrassRevetmentWaveImpactValidator.WaveAngleImpactQwa;

            const string errorMessage = "WaveAngleImpactQwa must be in range [0, 1].";

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMin, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, 0.0 - ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 0.0);
            ValidatorAssertHelper.AssertValue(validationFunc, 0.0 + ValidatorAssertHelper.Epsilon);

            ValidatorAssertHelper.AssertValue(validationFunc, 1.0 - ValidatorAssertHelper.Epsilon);
            ValidatorAssertHelper.AssertValue(validationFunc, 1.0);
            ValidatorAssertHelper.AssertValue(validationFunc, 1.0 + ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMax, ValidationIssueType.Error, errorMessage);
        }

        [Test]
        public void WaveAngleImpactRwa_VariousScenarios_ExpectedValues()
        {
            Func<double, ValidationIssue> validationFunc = GrassRevetmentWaveImpactValidator.WaveAngleImpactRwa;

            const string errorMessage = "WaveAngleImpactRwa must be larger than 0.";

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMin, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(validationFunc, 0.0 - ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 0.0, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(validationFunc, 0.0 + ValidatorAssertHelper.Epsilon);

            ValidatorAssertHelper.AssertValue(validationFunc, ValidatorAssertHelper.DoubleMax);
        }

        [Test]
        public void UpperLimitLoadingAul_VariousScenarios_ExpectedValues()
        {
            ValidationIssue ValidationFunc(double upperLimitLoadingAul) =>
                GrassRevetmentWaveImpactValidator.UpperLimitLoadingAul(upperLimitLoadingAul, 2);

            const string errorMessage = "UpperLimitLoadingAul must be smaller than LowerLimitLoadingAll.";

            ValidatorAssertHelper.AssertValue(ValidationFunc, ValidatorAssertHelper.DoubleMin);

            ValidatorAssertHelper.AssertValue(ValidationFunc, 2.0 - ValidatorAssertHelper.Epsilon);
            ValidatorAssertHelper.AssertValue(ValidationFunc, 2.0, ValidationIssueType.Error, errorMessage);
            ValidatorAssertHelper.AssertValue(ValidationFunc, 2.0 + ValidatorAssertHelper.Epsilon, ValidationIssueType.Error, errorMessage);

            ValidatorAssertHelper.AssertValue(ValidationFunc, ValidatorAssertHelper.DoubleMax, ValidationIssueType.Error, errorMessage);
        }
    }
}